### PR TITLE
2019-10-22 | User Custom Modal / Signals / Generic Class Base Views

### PR DESCRIPTION
**Pull request by iamhaiderkhan**

### CODE REVIEW

File: templates/registration/login.html [hunk 1]
```diff
@@ -0,0 +1,22 @@
...
+{% include 'home.html' %}
+<div class="container">
+    <div class="col-md-4 m-auto">
+        <div class="text-center">
+            <h1 class="mt-5 mb-4 ">Login</h1>
+        </div>
+        <form class="mb-2"  method="post">
+            {% csrf_token %}
+            <table>
+                {{form |crispy}}
```
Comment (ziafazal) [inline templates/registration/login.html:16]: why we put `crispy` filter here?

File: user_profile/views.py [hunk 2]
```diff
@@ -0,0 +1,24 @@
...
+        form.save()
+        return super().form_valid(form)
+
+
+class HomeView(ListView):
+    model = UserProfile
+    template_name = "home.html"
+    paginate_by = 50
+
+    def get_context_data(self, **kwargs):
```
Comment (ziafazal) [inline user_profile/views.py]: why we need to override `get_context_data `?

File: user_profile/models.py [hunk 3]
```diff
@@ -0,0 +1,28 @@
...
+    occupation = models.CharField(max_length=500, blank=True, verbose_name="Your Occupation")
+    bio = models.TextField(max_length=500, blank=True, verbose_name="About (Bio)")
+    address = models.CharField(max_length=255, blank=True, verbose_name="Permanent Address")
+    contact_number = models.CharField(max_length=30, blank=True, verbose_name="Contact Number")
+    city = models.CharField(max_length=30, blank=True, verbose_name="Your Location (City)")
+    country = models.CharField(max_length=30, blank=True, verbose_name="Your Nationality (Country)")
+    created_at = models.DateTimeField(auto_now_add=True)
+    updated_at = models.DateTimeField(auto_now=True)
+
+    @receiver(post_save, sender=User)
```
Comment (ziafazal) [inline user_profile/models.py]: both of these signal receivers don't make sense to me. We probably use `post_save` signal to send user welcome email.

File: user_profile/migrations/0005_auto_20191022_1727.py [hunk 4]
```diff
@@ -0,0 +1,20 @@
+# Generated by Django 2.2.6 on 2019-10-22 17:27
+
+from django.conf import settings
```
Comment (ziafazal) [inline user_profile/migrations/0005_auto_20191022_1727.py]: squash migrations

File: user_profile/forms.py [hunk 5]
```diff
@@ -0,0 +1,34 @@
+from django import forms
+from django.core.validators import RegexValidator
+from .models import User, UserProfile
+from django.contrib.auth.hashers import make_password
+from .mixins import EmailDuplicationMixin
+
+
+class SignUpForm(EmailDuplicationMixin, forms.ModelForm):
```
Comment (ziafazal) [inline user_profile/forms.py]: We probably should have two forms one for user and other for user_profile.  Using single forms complicates things when you are going to update user's profile.

**Review by ziafazal — CHANGES_REQUESTED** [review]
@iamhaiderkhan reviewed.

File: templates/registration/login.html (hunk 1 above)
Comment (iamhaiderkhan) [reply templates/registration/login.html:16]: it's just for Django customs form markup to bootstrap4 markup. 

Ref: https://pypi.org/project/django-crispy-forms/

File: user_profile/views.py (hunk 2 above)
Comment (iamhaiderkhan) [reply user_profile/views.py]: removed.

File: user_profile/models.py (hunk 3 above)
Comment (iamhaiderkhan) [reply user_profile/models.py]: 1st signal receiver (**create_user_profile**): if the newly user-created via signup it receives user profile fields and creates a user profile for the newly created user. I'll add sending welcome functionally this signal. 

2nd signal receiver (**save_user_profile**): if the user updated then the user profile update its updated_at field via user profile instance.

File: user_profile/forms.py (hunk 5 above)
Comment (iamhaiderkhan) [reply user_profile/forms.py]: Should I change signup into separate forms or handle update user's profiles with to different forms?

Comment (ziafazal) [reply user_profile/forms.py]: I think for both signup and update profile use cases.

Comment (iamhaiderkhan) [reply user_profile/forms.py]: Separate forms are created for User and UserProfile. SigUp and UpdateProfile both cases are handled.

File: user_profile/migrations/0005_auto_20191022_1727.py (hunk 4 above)
Comment (iamhaiderkhan) [reply user_profile/migrations/0005_auto_20191022_1727.py]: Done

File: user_profile/models.py (hunk 3 above)
Comment (iamhaiderkhan) [reply user_profile/models.py]: For now, I've removed both receivers. I'll add a post_save signal for sending a welcome email to newly created User in the next PR.

**Review by ziafazal — APPROVED** [review]
👍

**Pull request merged**